block-node/s3-archive/src/main/java/org/hiero/block/node/archive/s3/S3ArchivePlugin.java

```diff
@@ -177,104 +214,140 @@ public void handlePersisted(final PersistedNotification notification) {
-        final Iterator<BlockAccessor> blocksIterator = LongStream.range(startBlockNumber, endBlockNumber + 1)
-                .mapToObj(blockNumber -> context.historicalBlockProvider().block(blockNumber))
-                .iterator();
-        final Iterator<byte[]> tarBlocks = new TaredBlockIterator(Format.ZSTD_PROTOBUF, blocksIterator);
-        // fetch the first blocks consensus time so that we can place the file in a directory based on year and month
-        final BlockUnparsed firstBlock =
-                context.historicalBlockProvider().block(startBlockNumber).blockUnparsed();
-        final ZonedDateTime firstBlockConsensusTime;
-        try {
-            final Bytes headerBytes = firstBlock.blockItems().getFirst().blockHeader();
-            if (headerBytes == null) {
-                throw new IllegalStateException("Block header is null");
-            }
-            final BlockHeader header = BlockHeader.PROTOBUF.parse(headerBytes);
-            if (header.blockTimestamp() == null) {
-                throw new IllegalStateException("Block header firstTransactionConsensusTime is null");
+    private static S3Client createClient(final S3ArchiveConfig config) throws S3ClientInitializationException {
+        return new S3Client(
+                config.regionName(), config.endpointUrl(), config.bucketName(), config.accessKey(), config.secretKey());
+    }
+
+    /**
+     * An S3 upload task. This task uploads a
+     */
+    private static final class UploadTask implements Callable<Void> {
+        private static final System.Logger LOGGER = System.getLogger(UploadTask.class.getName());
+        private final long nextBatchStartBlockNumber;
+        private final long nextBatchEndBlockNumber;
+        private final S3ArchiveConfig archiveConfig;
+        private final AtomicLong lastArchivedBlockNumber;
+        private final BlockNodeContext context;
+
+        private UploadTask(
+                final long nextBatchStartBlockNumber,
+                final long nextBatchEndBlockNumber,
+                final AtomicLong lastArchivedBlockNumber,
+                final S3ArchiveConfig archiveConfig,
+                final BlockNodeContext context) {
+            this.nextBatchStartBlockNumber = nextBatchStartBlockNumber;
+            this.nextBatchEndBlockNumber = nextBatchEndBlockNumber;
+            this.archiveConfig = archiveConfig;
+            this.lastArchivedBlockNumber = lastArchivedBlockNumber;
+            this.context = context;
+        }
+
+        private static final String START_UPLOAD_MESSAGE = "Uploading archive block batch: {0} - {1}";
+        private static final String END_UPLOAD_MESSAGE = "Finished uploading archive block batch: {0} - {1}";
+
+        @Override
+        public Void call() throws S3ClientInitializationException, S3ResponseException, IOException {
+            // Check the nextBatchStartBlockNumber is less than lastArchivedBlockNumber, ie is this a duplicate
```

Comment:
nit: comments read in the opposite from the code.
Could have said somehting like
```suggestion
            // Check the nextBatchStartBlockNumber is greater than lastArchivedBlockNumber before attempting batched upload. If not we assume this is a duplicate and skip attempting upload for this batch as it may already be archived. This can also happen when the archive job has not been completed for a batch and a new batch is created
```